### PR TITLE
WindowTransform decrease amount of virtual function calls

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
WindowTransform decrease amount of virtual function calls.

### CODE REVIEW

**Comment by kitaisreal** [comment]
Prepare data:
```
CREATE TABLE test_table (id UInt64, value UInt64) ENGINE=MergeTree ORDER BY id;
INSERT INTO test_table SELECT number, number FROM numbers_mt(150000000);
```

Before:
```
SELECT id, sum(value) OVER test_window, min(value) OVER test_window, max(value) OVER test_window, avg(value) OVER test_window FROM test_table WINDOW test_window AS (PARTITION BY id % 25) FORMAT Null

0 rows in set. Elapsed: 9.767 sec. Processed 150.00 million rows, 2.40 GB (15.36 million rows/s., 245.74 MB/s.)
Peak memory usage: 2.69 GiB.

```
After:
```
SELECT id, sum(value) OVER test_window, min(value) OVER test_window, max(value) OVER test_window, avg(value) OVER test_window FROM test_table WINDOW test_window AS (PARTITION BY id % 25) FORMAT Null

0 rows in set. Elapsed: 8.429 sec. Processed 150.00 million rows, 2.40 GB (17.80 million rows/s., 284.74 MB/s.)
Peak memory usage: 2.67 GiB.
```

Before:
```
SELECT id, sum(value) OVER test_window, min(value) OVER test_window, max(value) OVER test_window, avg(value) OVER test_window FROM test_table WINDOW test_window AS (PARTITION BY id % 5) FORMAT Null

0 rows in set. Elapsed: 9.784 sec. Processed 150.00 million rows, 2.40 GB (15.33 million rows/s., 245.31 MB/s.)
Peak memory usage: 3.76 GiB.
```
After:
```
SELECT id, sum(value) OVER test_window, min(value) OVER test_window, max(value) OVER test_window, avg(value) OVER test_window FROM test_table WINDOW test_window AS (PARTITION BY id % 5) FORMAT Null

0 rows in set. Elapsed: 8.456 sec. Processed 150.00 million rows, 2.40 GB (17.74 million rows/s., 283.84 MB/s.)
Peak memory usage: 3.77 GiB.
```

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 4f448c87dbf3dcf0bee2570fc2f9bc52f17f1de9 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56120/4f448c87dbf3dcf0bee2570fc2f9bc52f17f1de9/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6691691238">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56120/4f448c87dbf3dcf0bee2570fc2f9bc52f17f1de9/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Review by nickitat — APPROVED** [review]

File: src/Processors/Transforms/WindowTransform.cpp [hunk 1]
```diff
@@ -987,9 +985,16 @@ void WindowTransform::writeOutCurrentRow()
             // FIXME does it also allocate the result on the arena?
             // We'll have to pass it out with blocks then...
 
-            /// We should use insertMergeResultInto to insert result into ColumnAggregateFunction
-            /// correctly if result contains AggregateFunction's states
-            a->insertMergeResultInto(buf, *result_column, arena.get());
+            if (ws.is_aggregate_function_state)
```
Comment (Algunenano) [inline src/Processors/Transforms/WindowTransform.cpp:988]: I guess a similar improvement might be possible in other places where `insertMergeResultInto` is called in a non constexpr environment right? I see FunctionInitializeAggregation and FunctionArrayReduce

**Review by Algunenano — APPROVED** [review]

File: src/Processors/Transforms/WindowTransform.cpp (hunk 1 above)
Comment (Algunenano) [reply src/Processors/Transforms/WindowTransform.cpp:988]: Adding a commit improving:
* FunctionInitializeAggregation -> `Select initializeAggregation('uniqExact', number) from numbers(100_000_000) settings max_threads=1 format Null;` around 20% faster
* ArrayReduce: Less improvement, around 10% on a query like: `Select arrayReduce('sum', a) from a settings max_threads=1 format Null;`

@kitaisreal WDYT?

Comment (kitaisreal) [reply src/Processors/Transforms/WindowTransform.cpp:988]: Yes, make sense. We need to merge this.

**Comment by Algunenano** [comment]
`01572_kill_window_function` became flaky because the query used there became thousands of times faster (from not finishing after 50 seconds when I got tired of waiting to finishing in 50ms). It should be better now

**Comment by Algunenano** [comment]
* Integration tests (tsan) [4/6] — fail: 1, passed: 421 -> `test_system_clusters_actual_information/test.py::test` -> https://github.com/ClickHouse/ClickHouse/issues/55168
* Integration tests (tsan) [5/6] — fail: 1, passed: 542 -> `test_ttl_move/test.py::test_alter_with_merge_work[mt_work]` -> https://github.com/ClickHouse/ClickHouse/issues/56686
* Stateless tests (aarch64) — fail: 1, passed: 5822, skipped: 30 -> `00002_log_and_exception_messages_formatting`. Currently flaky too
* Stress test (debug) — Check timeout expired -> The server took too long to shutdown. Not sure why

**Pull request merged**